tests/functional/autoprompt/test_prompttoolkit.py

```diff
@@ -51,45 +57,136 @@ def run(self, pre_run):
+                        return
+            if not self._queue:
+                app.after_render = Event(app, None)
+                app.exit()
+
+        self._app.after_render = Event(self._app, callback)
+
+        _stdout = io.TextIOWrapper(io.BytesIO(), encoding="utf-8")
+        self._app.output.stdout = _stdout
+        self._app.run(pre_run=pre_run)
+
+
+class BasicPromptToolkitTest(unittest.TestCase):
     @classmethod
     def setUpClass(cls):
-        cls.temporary_directory = tempfile.mkdtemp()
+        cls.test_file_creator = FileCreator()
         basename = 'tmpfile-%s' % str(random_chars(8))
-        full_filename = os.path.join(cls.temporary_directory, basename)
+        full_filename = cls.test_file_creator.full_path(basename)
         _generate_index(full_filename)
         cls.completion_source = create_autocompleter(
             full_filename, response_filter=filters.fuzzy_filter)
 
+        history = {
+            'version': 1,
+            'commands': [
+                'accessanalyzer update-findings',
+                'ec2 describe-instances',
+                's3 ls'
+            ]
+        }
+        cls.history_filename = cls.test_file_creator.create_file(
+            'prompt_history.json', json.dumps(history))
+
     @classmethod
     def tearDownClass(cls):
-        shutil.rmtree(cls.temporary_directory)
+        cls.test_file_creator.remove_all()
 
     def setUp(self):
         self.completer = PromptToolkitCompleter(self.completion_source)
-        self.factory = PromptToolkitFactory(self.completer)
+        self.history_driver = HistoryDriver(self.history_filename)
         self.driver = create_clidriver()
-        self.app = self.create_application()
+        self.factory = PromptToolkitFactory(
+            self.completer,
+            history_driver=self.history_driver
+        )
         self.prompter = PromptToolkitPrompter(
-            completion_source=self.completion_source, driver=self.driver,
-            app=self.app)
+            completion_source=self.completion_source,
+            driver=self.driver,
+            factory=self.factory
+        )
+        self.prompter.args = []
         self.prompter.input_buffer = self.factory.create_input_buffer()
         self.prompter.doc_buffer = self.factory.create_doc_buffer()
 
+    def create_application(self):
+        layout = self.factory.create_layout()
+        return Application(layout=layout)
+
+    def add_current_buffer_assertion(self, buffer_name):
+        return lambda app: app.current_buffer.name == buffer_name
+
+    def add_buffer_is_visible_assertion(self, buffer_name):
+        return lambda app: buffer_name in [window.content.buffer.name
+                    for window in app.layout.get_visible_focusable_windows()]
+
+    def add_buffer_not_visible_assertion(self, buffer_name):
+        return lambda app: buffer_name not in [window.content.buffer.name
+                    for window in app.layout.get_visible_focusable_windows()]
```

Comment:
These look good, but let's rename them to use `get` instead of `add`. I'm mainly suggesting this because when comparing it to the `add_keypress` and `add_keypress_with_assertion` these methods are only generate the assertions but do not affect test results because they are not added to the action queue.